File: src/services/openaiService.js
const OpenAI = require('openai');
require('dotenv').config();

class OpenAIService {
    constructor() {
        if (!process.env.OPENAI_API_KEY) {
            throw new Error('OPENAI_API_KEY nie je nastavený v .env súbore');
        }

        this.openai = new OpenAI({
            apiKey: process.env.OPENAI_API_KEY
        });
    }

    async validateQuestionSelection(selectedQuestion, previousAnswers) {
        try {
            // Najprv otestujeme API kľúč
            await this.testAPIKey();
            
            const prompt = this.createValidationPrompt(selectedQuestion, previousAnswers);
            
            const response = await this.openai.chat.completions.create({
                model: "gpt-4",
                messages: [
                    {
                        role: "system",
                        content: "Si expert na výber otázok pre darčekový poradca. Tvoja úloha je validovať, či je vybraná otázka relevantná a užitočná pre získanie informácií o preferenciách pre darček."
                    },
                    {
                        role: "user",
                        content: prompt
                    }
                ],
                temperature: 0.7,
                max_tokens: 150
            });

            const validation = response.choices[0].message.content;
            return this.parseValidationResponse(validation);
        } catch (error) {
            console.error('Chyba pri validácii otázky s OpenAI:', error);
            if (error.response?.status === 401) {
                console.error('Neplatný OpenAI API kľúč. Skontrolujte .env súbor.');
            }
            return false;
        }
    }

    async testAPIKey() {
        try {
            await this.openai.chat.completions.create({
                model: "gpt-4",
                messages: [{ role: "user", content: "Test" }],
                max_tokens: 5
            });
            console.log('OpenAI API kľúč je platný');
            return true;
        } catch (error) {
            if (error.response?.status === 401) {
                throw new Error('Neplatný OpenAI API kľúč');
            }
            throw error;
        }
    }

    createValidationPrompt(selectedQuestion, previousAnswers) {
        return `
        Predchádzajúce odpovede: ${JSON.stringify(previousAnswers)}
        Vybraná otázka: ${JSON.stringify(selectedQuestion)}
        
        Je táto otázka relevantná a užitočná pre získanie informácií o preferenciách pre darček?
        Odpovedz len "ÁNO" alebo "NIE".
        `;
    }

    parseValidationResponse(response) {
        return response.trim().toUpperCase() === 'ÁNO';
    }
}

module.exports = OpenAIService; 

File: src/services/openai_service.py
import os
from openai import OpenAI
from dotenv import load_dotenv

class OpenAIService:
    def __init__(self):
        load_dotenv()
        if not os.getenv('OPENAI_API_KEY'):
            raise ValueError('OPENAI_API_KEY nie je nastavený v .env súbore')
        self.client = OpenAI(api_key=os.getenv('OPENAI_API_KEY'))

    def test_api_key(self):
        try:
            self.client.chat.completions.create(
                model="gpt-3.5-turbo",
                messages=[{"role": "user", "content": "Test"}]
            )
            print('OpenAI API kľúč je platný')
            return True
        except Exception as e:
            print(f'Chyba pri testovaní OpenAI API kľúča: {str(e)}')
            return False

    def validate_question_selection(self, selected_question, previous_answers):
        try:
            # Vytvoríme prompt pre validáciu
            prompt = self.create_validation_prompt(selected_question, previous_answers)
            
            # Získame odpoveď od OpenAI
            response = self.client.chat.completions.create(
                model="gpt-3.5-turbo",
                messages=[{"role": "user", "content": prompt}]
            )
            
            # Spracujeme odpoveď
            return self.parse_validation_response(response.choices[0].message.content)
        except Exception as e:
            print(f'Chyba pri validácii otázky: {str(e)}')
            return False

    def create_validation_prompt(self, selected_question, previous_answers):
        prompt = f"Validuj, či je nasledujúca otázka relevantná na základe predchádzajúcich odpovedí:\n\n"
        prompt += f"Predchádzajúce odpovede:\n"
        for i, answer in enumerate(previous_answers, 1):
            prompt += f"{i}. {answer}\n"
        prompt += f"\nVybraná otázka: {selected_question['text']}\n"
        prompt += "\nOdpovedz len 'áno' alebo 'nie'."
        return prompt

    def parse_validation_response(self, response):
        response = response.lower().strip()
        return response in ['áno', 'yes', 'true', '1'] 